File: backend/src/main/java/br/ufpr/sept/so2/modules/iam/infrastructure/security/RefreshCookieFactory.java
package br.ufpr.sept.so2.modules.iam.infrastructure.security;

import br.ufpr.sept.so2.modules.iam.infrastructure.IamProperties;
import org.springframework.http.ResponseCookie;
import org.springframework.stereotype.Component;

import java.time.Duration;

@Component
public class RefreshCookieFactory {

    private final IamProperties properties;

    public RefreshCookieFactory(IamProperties properties) {
        this.properties = properties;
    }

    public ResponseCookie create(String refreshToken) {
        return base()
                .value(refreshToken)
                .maxAge(Duration.ofSeconds(properties.refreshTtlSeconds()))
                .build();
    }

    public ResponseCookie clear() {
        return base()
                .value("")
                .maxAge(Duration.ZERO)
                .build();
    }

    public String cookieName() {
        return properties.getCookieName();
    }

    private ResponseCookie.ResponseCookieBuilder base() {
        return ResponseCookie.from(properties.getCookieName(), "")
                .httpOnly(true)
                .secure(properties.isCookieSecure())
                .sameSite(properties.getCookieSameSite())
                .path("/auth");
    }
}

File: backend/src/main/kotlin/br/ufpr/sept/so2/modules/iam/infrastructure/security/RefreshCookieFactory.kt
package br.ufpr.sept.so2.modules.iam.infrastructure.security

import br.ufpr.sept.so2.modules.iam.infrastructure.IamProperties
import org.springframework.http.ResponseCookie
import org.springframework.stereotype.Component
import java.time.Duration

@Component
class RefreshCookieFactory(
    private val properties: IamProperties,
) {
    fun create(refreshToken: String): ResponseCookie =
        base()
            .value(refreshToken)
            .maxAge(Duration.ofSeconds(properties.refreshTtlSeconds))
            .build()

    fun clear(): ResponseCookie =
        base()
            .value("")
            .maxAge(Duration.ZERO)
            .build()

    fun cookieName(): String = properties.cookieName

    private fun base(): ResponseCookie.ResponseCookieBuilder =
        ResponseCookie.from(properties.cookieName, "")
            .httpOnly(true)
            .secure(properties.cookieSecure)
            .sameSite(properties.cookieSameSite)
            .path("/auth")
}
